### PR TITLE
SQLStore: Create utility for bulk inserts

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

This PR is to support https://github.com/grafana/grafana/pull/56575, along with some other similar use cases in Alerting.

This PR implements a small utility over the ORM which does bulk inserts, that splits large enough inserts into chunks. The batch size is configurable, and we give a way to pick reasonable defaults based on the dialect being used.

**Why do we need this feature?**

xorm's `InsertMulti()` and other bulk-style family of functions tend to dump all requested operations into a single SQL statement. In practice, we almost always want batching, instead. Batching avoids sending overly large single statements, reduces lock and disk contention, and works around limitations in databases - namely, caps on the number of parameters allowed in a single statement.

**Who is this feature for?**

Initially, this feature will be used in 3 places in the Alerting subsystem where we have potentially large inserts - annotations, state persistence, and the migration from Legacy alerting.

It can also be used as an alternative to `InsertMulti`, by any other subsystem that needs it, for a bit more robustness when dealing with large operations.

**Which issue(s) does this PR fix?**:

Contributes to: https://github.com/grafana/grafana/issues/54096

**Special notes for your reviewer**:

n/a

### CODE REVIEW

File: pkg/services/sqlstore/bulk.go
Comment (alexweav) [inline pkg/services/sqlstore/bulk.go:16]: I'd have liked to hide this detail, but I really wanted to make this work at the session level rather than the store level. This is so that the caller has complete control over where this utility is used - we will respect transactions or any other context on the same session.

Unfortunately, the exported API for session doesn't seem to offer a way to obtain the dialect of its parent store. It seemed risky to add such a link here.

File: pkg/services/sqlstore/bulk.go
Comment (alexweav) [inline pkg/services/sqlstore/bulk.go:29]: Accepting `interface{}` for the records is a necessary evil:
- It's consistent with xorm's API. This reflection-based approach is what xorm already does under the hood.
- Slices aren't covariant in Go, so accepting a `[]interface{}` forces the caller to convert from the type they are using, which is a risk for unneeded copies. Here, I focused on minimizing copies.
- Generics are an _okay_ solution, but can't be used on methods. The method approach offered a cleaner API, was more discoverable, and feels more native to current xorm users.

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM, nice addition to our tooling 👍

**Review by yuri-tceretian — COMMENTED** [review]
The changes LGTM but I would like to have formal approval from the platform team

**Comment by mildwonkey** [comment]
👋 Hi! This looks reasonable enough and worth trying in my opinion, but I'd also prefer to tag in someone with more actual sql/xorm experience. I'll make sure the rest of the platform squad sees this.

**Comment by zserge** [comment]
Not bad in general, but as we hope to move to SQLx sessions eventually - it would be nice to see a SQLx session implementation of this before merging.

File: pkg/services/sqlstore/bulk.go
Comment (papagian) [inline pkg/services/sqlstore/bulk.go:24]: some debug message would be useful

**Review by papagian — APPROVED** [review]
LGTM; great work

**Comment by alexweav** [comment]
I may hold off on the SQLx implementation for two reasons:
1. I'm not an expert with SQLx, and I'm not certain that this API or approach makes sense in that context. It's not clear to me how connections/transactions are handled there, or what a native API would "feel" like. I assume it'd be different, as this PR tries to match the feel of xorm.
2. The InBatches implementation is exported and is database-agnostic, so SQLx users can still leverage the most important part as-is.

**Comment by alexweav** [comment]
CodeQL step seems to be hanging across the board today. It's ran for an hour and a half now, just gonna merge.

**Pull request merged**
